FIX #2411 On a btrfs system, don't default to aufs

robvanmieghem opened the pull request:
FIX #2411 On an Ubuntu system, don't default to aufs on a btrfs filesystem

Signed-off-by: Rob Van Mieghem <rob@vanmieghemcloud.com>

nathanleclaire commented:
Seems pretty reasonable, although I wish we had better test coverage for the provisioners so we could add some for this case.

nathanleclaire commented:
If you're up to the challenge, it would be really great to get some unit tests for at least some of these code changes.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "master" git@github.com:robvanmieghem/machine.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

robvanmieghem commented:
@nathanleclaire you are absolutely right. Just added them.

nathanleclaire commented:
Nice thanks for adding the tests I have some feedback on the code but I am willing to add this feature once it gets addressed.  cc @docker/machine-maintainers also for review.

nathanleclaire commented on libmachine/provision/ubuntu_systemd_test.go:
```diff
@@ -27,3 +32,34 @@ func TestUbuntuSystemdCompatibleWithHost(t *testing.T) {
+
+func TestUbuntuSystemdStorageDriver(t *testing.T) {
+
+	p := NewUbuntuSystemdProvisioner(&fakedriver.Driver{}).(*UbuntuSystemdProvisioner)
```
Can you please convert this and the other test to use a [test table](https://github.com/golang/go/wiki/TableDrivenTests)?  It seems to be a good fit for what you're doing here, and there are a couple other examples of doing this across the codebase (e.g. `commands/env_test.go`) if you want to take a look.

nathanleclaire commented on libmachine/provision/utils.go:
```diff
@@ -229,3 +238,17 @@ func waitForDocker(p Provisioner, dockerPort int) error {
+
+//fakeSSHCommander is intended to be used by testing code
+//Extend it when needed
+type fakeSSHCommander struct {
```
Can you please make this exported (i.e. named `FakeSSHCommander` instead of `fakeSSHCommander`) and move it's definition into a test with `_test.go` name to delineate it from non-test code?  One suggestion is that you could add a `provisiontest` module inside of `provision` module (like how the standard library has `httptest`) and put it in a file there.

nathanleclaire commented on libmachine/provision/utils.go:
```diff
@@ -208,6 +209,14 @@ func matchNetstatOut(reDaemonListening, netstatOut string) bool {
 	return false
 }
 
+func getFilesystemType(p Provisioner, directory string) (fstype string, err error) {
```
Nit: Can you please change the named return parameters to being explicitly defined in the function body?  That's convention in this codebase.

nathanleclaire commented on libmachine/provision/ubuntu_upstart.go:
```diff
@@ -135,14 +135,30 @@ func (provisioner *UbuntuProvisioner) dockerDaemonResponding() bool {
 	return true
 }
 
+func (provisioner *UbuntuProvisioner) setStorageDriver() error {
```
Instead of duplicating this function for the two provisioners, why not make a helper function that you invoke from `Provision` method instead?  e.g. the invocation would look something like (in `Provision` method):

```golang
storageDriver, err := decideStorageDriver(provisioner, provisioner.EngineOptions.StorageDriver)
if err != nil {
    return err
}
```

nathanleclaire commented:
Thanks for your contribution @robvanmieghem! Left some comments on the code.

As a side note, I'd really like to switch the magic `"aufs"`, `"btrfs"`, etc. constants to use real `const`, but I think it's best addressed in a followup PR (since it was already that way before and is out of scope of this change).

robvanmieghem commented:
Thank you for reviewing and commenting on this pull request @nathanleclaire !

I modified the code in response to your notes. Since the storagedriver decision is now extracted from the Ubuntu provisioners, I added this this method call to the provisioners that set a default storagedriver as well. This should effectively only impact Debian since it uses aufs as a default too.

You might notice that I introduced a fakeProvisioner in the utils_test.go file. It's not in the `provisiontest` package since implementing the provisioner interface would cause a circular package dependency since this code would be used by the tests inside the `provision` package. 

I agree that refactoring these the magic "aufs", "btrfs", etc. constants to real consts is an easy improvement but might get this PR to get out of it's initial scope.

nathanleclaire commented on libmachine/provision/provisiontest/sshcommander.go:
```diff
@@ -0,0 +1,22 @@
+//Extend it when needed
+type FakeSSHCommander struct {
+	//Result of the ssh command to look up the FilesystemType
+	FilesystemType string
```
Hm, instead of having a field specifically for this, could you maybe make some sort of lookup table field which is `map[string]string`?  Then you could define the result of a given call like so:

```console
cmder := &FakeSSHCommander{
    Responses: map[string]string{
        "stat -f -c %T /var/lib": "btrfs\n",
    },
}
```

WDYT?

robvanmieghem replied:
This would indeed be perfect for testing the output parsing.
For testing more high level functionality, like checking the default graphdriver in the RedHat provisioner for example, it does not really make sense to know that the underlying call is `stat -f -c %T /var/lib`, this would also prevent easy refactoring of the specific implementation of the filesystem checking without modifying all tests.
Both kind of tests are necessary so I'll refactor the FakeSSHCommander to use an internal command/resonse map for output parsing tests while a creationmethod with more high level configuration can be used by the other ones.

nathanleclaire replied:
> For testing more high level functionality, like checking the default graphdriver in the RedHat provisioner for example, it does not really make sense to know that the underlying call is stat -f -c %T /var/lib, this would also prevent easy refactoring of the specific implementation of the filesystem checking without modifying all tests.

I'm not sure I understand.  If the underlying command sent to SSH changes (say, for Red Hat), I _want_ the tests to break.  We'll then know that the tests need to be updated as well, will have clearly defined expected behavior, and will only need to update one line in the tests.

nathanleclaire commented:
Nice everything is looking pretty good on my end, I posted one more comment with an additional ask.

@docker/machine-maintainers Please take a look at this and let us know your thoughts when you get a chance.

dgageot commented:
LGTM

jeanlaurent commented:
LGTM 

@nathanleclaire I let you give the final go/merge.

nathanleclaire commented:
I thought of something else as well.  If user specifies something like `--engine-opt graph=/vol/docker`, the daemon will create and use the graph directory at `/vol/docker` instead of `/var/lib/docker`.  So, we should probably account for this case as well.

nathanleclaire commented:
Consequently, you should most likely pass in the entire `EngineOptions` to `decideStorageDriver`, loop to check if this set in `ArbitraryFlags` (`if strings.HasPrefix(flag, "graph=")`) and over-ride the default value of `/var/lib` if so.  It will require a bit of [path](https://golang.org/pkg/path/) parsing, most likely.

Maybe we should open another ticket and address it in a follow-up PR, but I wanted to make sure everyone is aware of the issue and a potential solution.

robvanmieghem commented:
@nathanleclaire , I modified the FakeSSHCommander as explained earlier. I hope this makes it a bit more clear and both ways of testing are supported this way. If you want to have it ammended in the previous commit, just let me know. Thought a separate commit makes the FakeSSHCommander for testing more clear.

robvanmieghem commented:
>I thought of something else as well. If user specifies something like --engine-opt graph=/vol/docker, the daemon will create and use the graph directory at /vol/docker instead of /var/lib/docker. So, we should probably account for this case as well.
You are right, did not think of that...

nathanleclaire commented:
> You are right, did not think of that...

Like I said, maybe we can just create a ticket to track the issue, and merge this (since I'm not aware of too many folks setting custom graph drivers today).  Then follow up later.

nathanleclaire commented:
LGTM
